qir/qat/RecordPullBackPass/RecordPullBackPass.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "RecordPullBackPass.hpp"
+
+#include "Logging/ILogger.hpp"
+#include "QatTypes/QatTypes.hpp"
+#include "RecordPullBackPass/RecordPullBackPass.hpp"
+
+#include "Llvm/Llvm.hpp"
+
+#include <functional>
+#include <stdexcept>
+#include <unordered_map>
+#include <vector>
+
+namespace microsoft::quantum
+{
+
+std::string const RecordPullBackPass::RECORD_INSTR_END = "_record_output";
+
+llvm::PreservedAnalyses RecordPullBackPass::run(llvm::Function& function, llvm::FunctionAnalysisManager& /*mam*/)
```

Comment:
There are two issues with this pass: 

1) The immediate need is to ensure that all quantum instructions that measurements are followed only be other measurements, record output functions (and the ret void), when the initial IR mixes measurements and other __quantum__qis* calls. You can rely on only record output functions depending on measurement results and nothing else. 

2) Not relevant to solve the immediate need 1, but general feed back on this logic: I believe we still have an issue with the logic separation not recognizing record output function. Running this pass after logic separation does not address the need since the logic separation distroys the order of the record output functions.